### DISCUSSION TITLE
Using bare-bones zsim for Z80 developing

### DISCUSSION THREAD

**Discussion by ewkohler** [maziac/DeZog#112] [Q&A] [answered]
Hello, I am trying to configure VSCode on Windows with DeZog for programming algorithms in Z80 (no ZX Spectrum emulation). I tried to create simple .asm, launch.json and tasks.json. However, when starting debugging, I am able to step the commands, but the Registers view is not updated. (only PC). When finished, VSCode opens ‘disasm.lst’, but I can’t see any code in there. Can anyone help? Thanks.

launch.json:
`{
    "configurations": [
    {
        "type": "dezog",
        "request": "launch",
        "name": "Simulator - Bare Z80",
        "remoteType": "zsim",
        "zsim": {
            "visualMemory": true,
            "memoryModel": "RAM"
        },
        "sjasmplus": [
            {
                "path": "main.sld"
            }
        ],
        "commandsAfterLaunch": [],
        "history": {
            "reverseDebugInstructionCount": 1000000,
            "spotCount": 10,
            "codeCoverageEnabled": true
        },
        "startAutomatically": false,
        "rootFolder": "${workspaceFolder}",
        // "loadObjs": [
        //     {
        //         "path": "The .obj file to load"
        //    }
        //],
        "execAddress": "main",
        "topOfStack": "stack_top"
    }
    ]
}`

tasks.json
`{
    "version": "2.0.0",
    "tasks": [
        {
            "label": "make (sjasmplus)",
            "type": "shell",
            "command": "sjasmplus", // Please add the full path to sjasmplus
            "args": [
                "--sld=main.sld",
                "--fullpath",
                "main.asm"
            ],
            "problemMatcher": {
                "owner": "sjasmplus",
                "fileLocation": "autoDetect",
                "pattern": {
                    "regexp": "^(.*)\\((\\d+)\\):\\s+(warning|error):\\s+(.*)$",
                    "file": 1,
                    "line": 2,
                    "severity": 3,
                    "message": 4
                }
            },
            "group": {
                "kind": "build",
                "isDefault": true
            }
        }
    ]
}
`

https://user-images.githubusercontent.com/34894226/229681793-905f6a83-0c19-414f-9694-95899dd39500.mp4

**Comment by maziac** [comment]
Could you ZIP and put the whole project here with list (or sld) File.

**Comment by ewkohler** [comment]
[z80-programming-exercicios.zip](https://github.com/maziac/DeZog/files/11148403/z80-programming-exercicios.zip)
Sure, here it is. Thanks!

**Comment by maziac** [comment, marked as answer]
You didn't load any binary to the simulator:
![Bildschirmfoto 2023-04-04 um 16 46 09](https://user-images.githubusercontent.com/3081973/229830014-0be4bcc4-765d-4162-8889-8e264a60216c.jpg)
In the VARIABLES pane you can see the "real" assembly code. It is all NOPs.
You can step through the nops but no register will change.
You see the source code stepping work because the debugger simply associates the addresses with your source code.
And for $8000 and so forth there are valid addresses in your source code.

Here is what you have to do:
1. sjasmplus currently does not produce any binary code. E.g .add 
				"--raw=main.bin"
to your tasks.json (args).
2. In your launch.json instruct dezog to load your binary. This is done with:
~~~
			"loadObjs": [
				{
					"path": "main.bin",
					"start": "main"
				}
			],
~~~
"start":"main"  is the start address of your binary, i.e. where to load it.
3. Start the debugger.
4. To verify, open the Disassembly in VARIABLES.
Should now be:
![Bildschirmfoto 2023-04-04 um 17 03 19](https://user-images.githubusercontent.com/3081973/229835100-c1bc9092-fd84-4dda-acbd-b3b88f467d65.jpg)

**Reply by ewkohler** [reply]
Everything worked except that in loadObjs/start I had to input the actual address ($8000) instead of 'main', as your example. When I used "start": "main" I got the error "cannot evaluate 'loadObjs[].start' (main).". Is this a problem or the way it works? Can I use the label somehow instead of the actual start address? Thanks!

![image](https://user-images.githubusercontent.com/34894226/229915020-149c2215-b881-4bcd-bfed-1b78155e58c0.png)

Working 'loadObjs':
```
        "loadObjs": [
             {
                 "path": "main.bin",
                 "start": "$8000"
            }
        ],
```

**Reply by maziac** [reply]
A label should work (and worked for me).
Which version of dezog are you using?

**Reply by ewkohler** [reply]
![image](https://user-images.githubusercontent.com/34894226/229928336-22ac7dc3-a02f-44e4-abeb-897fbc8c5758.png)
Version 3.2.4. Maybe some flag during compiling? I could only find absolute addresses in documentation...

**UPDATE**: if I launch using "main" it throws error. Change it to "$8000", run debug, stop, then change back to "main", and it works.

**Comment by maziac** [comment]
Sorry, I mixed up 'start' and 'execAddress'.
Yes, you are right: for 'start' the labels do not work yet.

It has to do with the order of processing. The bin file was loaded before the sld file was parsed.
I changed the order now.
Please try the attached build.
[dezog-3.2.5-rc1.vsix.zip](https://github.com/maziac/DeZog/files/11157086/dezog-3.2.5-rc1.vsix.zip)

**Comment by maziac** [comment]
Although I had the change in 2.3.5-rc1 I need to revert it.
The problem is with zesarux.
Reading the sld file requires that the memory model of the target is known. 
zesarux might change the memory model on loading.
I.e. I have to read the labels after loading, unfortunately.

I could, of course, use different approaches for the different remotes, but I would rather like to keep it similar over all remotes.

So, for 'start' you cannot use a label, but you have to use a number.

**Reply by ewkohler** [reply]
Ok, no problem. Thanks very much for your help and amazing job!

**Comment by maziac** [comment]
I found a way to do it by separating 'load' and 'loadObjs'.
Maybe you can give this a try:
[dezog-3.2.5-rc2.vsix.zip](https://github.com/maziac/DeZog/files/11168407/dezog-3.2.5-rc2.vsix.zip)
'start' with using a label should work now.

**Comment by maziac** [comment]
Fixed in DeZog 3.3.0
